File: demos/php/algorithms_demo.php
<?php

function binarySearch(array $values, int $target): int {
    $low = 0;
    $high = count($values) - 1;
    while ($low <= $high) {
        $middle = intdiv($low + $high, 2);
        if ($values[$middle] === $target) return $middle;
        if ($values[$middle] < $target) $low = $middle + 1;
        else $high = $middle - 1;
    }
    return -1;
}

function bfs(array $graph, string $start): array {
    $order = [];
    $seen = [];
    $queue = new SplQueue();
    $queue->enqueue($start);
    while (!$queue->isEmpty()) {
        $node = $queue->dequeue();
        if (isset($seen[$node])) continue;
        $seen[$node] = true;
        $order[] = $node;
        foreach ($graph[$node] ?? [] as $next) $queue->enqueue($next);
    }
    return $order;
}

$scores = [['name' => 'Lin', 'score' => 91], ['name' => 'Ada', 'score' => 95], ['name' => 'Kai', 'score' => 91]];
usort($scores, fn ($a, $b) => $b['score'] <=> $a['score'] ?: $a['name'] <=> $b['name']);
$numbers = [2, 5, 8, 13, 21];
$graph = ['A' => ['B', 'C'], 'B' => ['D'], 'C' => ['D']];

echo 'sorted=' . implode(',', array_map(fn ($x) => "{$x['name']}:{$x['score']}", $scores)) . "\n";
echo 'binary-search-13=' . binarySearch($numbers, 13) . "\n";
echo 'bfs=' . implode(',', bfs($graph, 'A')) . "\n";

File: demos/go/algorithms_demo.go
package main

import (
	"fmt"
	"sort"
)

type Score struct {
	Name  string
	Value int
}

func bfs(graph map[string][]string, start string) []string {
	order := []string{}
	seen := map[string]bool{}
	queue := []string{start}
	for len(queue) > 0 {
		node := queue[0]
		queue = queue[1:]
		if seen[node] { continue }
		seen[node] = true
		order = append(order, node)
		queue = append(queue, graph[node]...)
	}
	return order
}

func main() {
	scores := []Score{{"Lin", 91}, {"Ada", 95}, {"Kai", 91}}
	sort.SliceStable(scores, func(i, j int) bool {
		if scores[i].Value != scores[j].Value { return scores[i].Value > scores[j].Value }
		return scores[i].Name < scores[j].Name
	})
	numbers := []int{2, 5, 8, 13, 21}
	index := sort.SearchInts(numbers, 13)
	graph := map[string][]string{"A": {"B", "C"}, "B": {"D"}, "C": {"D"}}

	fmt.Printf("sorted=%v\n", scores)
	fmt.Printf("binary-search-13=%d\n", index)
	fmt.Printf("bfs=%v\n", bfs(graph, "A"))
}
